File: 21-Binary-Search-Tree-Part-II/python/04_kth_smallest_and_largest_element_in_a_bst.py
from typing import List, Optional

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kLargesSmall(self, root: Optional[TreeNode], k: int) -> List[int]:
        inorder = []

        def inorder_traversal(node: Optional[TreeNode]):
            if not node:
                return
            inorder_traversal(node.left)
            inorder.append(node.val)
            inorder_traversal(node.right)

        inorder_traversal(root)

        return [inorder[k - 1], inorder[len(inorder) - k]]

File: 21-Binary-Search-Tree-Part-II/java/04-KthSmallestAndLargestElementInABst.java
import java.util.*;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> kLargesSmall(TreeNode root, int k) {
        List<Integer> inorder = new ArrayList<>();
        inorderTraversal(root, inorder);

        List<Integer> result = new ArrayList<>();
        result.add(inorder.get(k - 1));
        result.add(inorder.get(inorder.size() - k));

        return result;
    }

    private void inorderTraversal(TreeNode node, List<Integer> list) {
        if (node == null) return;
        inorderTraversal(node.left, list);
        list.add(node.val);
        inorderTraversal(node.right, list);
    }
}
